In [1]:
import numpy as np
from scipy.optimize import differential_evolution

In [4]:
def ackley(x):
    arg1 = -0.2 * np.sqrt(0.5 * (x[0] ** 2 + x[1] ** 2))
    arg2 = 0.5 * (np.cos(2. * np.pi * x[0]) + np.cos(2. * np.pi * x[1]))
    return -20. * np.exp(arg1) - np.exp(arg2) + 20. + np.e

bounds = [(-10, 10), (-10, 10)]

In [6]:
%%time
result = differential_evolution(ackley, bounds, seed=42)
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 2793
     nit: 90
 success: True
       x: array([0., 0.])
CPU times: user 244 ms, sys: 3.93 ms, total: 248 ms
Wall time: 249 ms


C параметрами по умолчанию поиск оптимального решения выполнился за:
- 249 минут
- кол-во операции в целом 2793
- кол-во вычислений в 1 итерации 90

Исследуемые праметры по умолчанию:
- strategy => 'best1bin'
- tol => 0.01
- maxiter => 1000
- popsize => 15
- mutation => (0.5,1.0)
- recombination  => 0.7
- updating => 'immediate'
- workers => 1 
 

In [45]:
%%time
result = differential_evolution(ackley, bounds, seed=42,updating ='deferred', 
                                workers = 1,maxiter = 100000,popsize = 50,
                               mutation = 1.0 ,tol = 0.0000000001)
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 18863
     nit: 187
 success: True
       x: array([0., 0.])
CPU times: user 1.1 s, sys: 7.15 ms, total: 1.11 s
Wall time: 1.12 s


In [46]:
%%time
result = differential_evolution(ackley, bounds, seed=42,updating ='deferred', 
                                workers = 4,maxiter = 100000,popsize = 50,
                               mutation = 1.0 ,tol = 0.0000000001)
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 18863
     nit: 187
 success: True
       x: array([0., 0.])
CPU times: user 1.2 s, sys: 19.7 ms, total: 1.22 s
Wall time: 1.34 s


#### Вывод по иследованию параметра updating
Изменения параметра updating с immediate на deferred с одной стороны замедляет работу алгоритма, с другой стороны дает возможность распараллелить его работу при помощи параментра workers > 1. Я установил остальные параметры, для того чтобы алоритм работал как можно дольше. Резульаты удивили. С паралельностью алоритм работает дольше чем в один поток. Вероятно это связано, с тем что работа алгоритма на столько не значительна по времени, что время затрачиваемое на организацию распораллеливания и последующим объединением резальтатов работы каждого потока,  существенно относительно времени работы самого алгоритма. Вероятно, на более ресурсоемких данных, пралельность будет иметь смысл. 

In [50]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 100000,popsize = 1,
                                mutation = 1.0 ,recombination = 0.7,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 3.5744518772578115
     jac: array([ 1.42108547e-06, -3.19744231e-06])
 message: 'Optimization terminated successfully.'
    nfev: 82
     nit: 10
 success: True
       x: array([-0.96847764,  0.96847759])
CPU times: user 10.9 ms, sys: 0 ns, total: 10.9 ms
Wall time: 9.68 ms


In [51]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 100000,popsize = 5,
                                mutation = 1.0 ,recombination = 0.7,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 1.9754271551875036e-08
     jac: array([-0.00076987,  0.05022187])
 message: 'Optimization terminated successfully.'
    nfev: 323
     nit: 19
 success: True
       x: array([-5.00190150e-09, -4.87441178e-09])
CPU times: user 25.3 ms, sys: 3.78 ms, total: 29.1 ms
Wall time: 29.8 ms


In [56]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 100000,popsize = 20,
                                mutation = 1.0 ,recombination = 0.7,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 6263
     nit: 154
 success: True
       x: array([0., 0.])
CPU times: user 536 ms, sys: 0 ns, total: 536 ms
Wall time: 537 ms


#### Вывод по иследованию параметра popsize
Вариация параметра popsize, дает прирост производительности вычисления, чем меньше параметр, тем быстрее работает алгоритм. Однако при этом страдает точность вычисления. При popsize=1 решение не удовлетворительное. При popsize=5 решение вполне приемлемое. При popsize=20 решение отличное. Координаты X определились без погрешности.

In [58]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 1000,popsize = 20,
                                mutation = 1.0 ,recombination = 0.7,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 6263
     nit: 154
 success: True
       x: array([0., 0.])
CPU times: user 543 ms, sys: 3.54 ms, total: 547 ms
Wall time: 547 ms


In [59]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 100,popsize = 20,
                                mutation = 1.0 ,recombination = 0.7,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 1.6757839560455068e-10
 message: 'Maximum number of iterations has been exceeded.'
    nfev: 4103
     nit: 100
 success: False
       x: array([ 5.38857847e-11, -2.46291876e-11])
CPU times: user 354 ms, sys: 0 ns, total: 354 ms
Wall time: 352 ms


In [57]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 10,popsize = 20,
                                mutation = 1.0 ,recombination = 0.7,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 2.0059755190260375e-08
     jac: array([ 0.01510614, -0.02704006])
 message: 'Maximum number of iterations has been exceeded.'
    nfev: 578
     nit: 10
 success: False
       x: array([-4.96197875e-09, -5.06734522e-09])
CPU times: user 43.5 ms, sys: 0 ns, total: 43.5 ms
Wall time: 43.2 ms


#### Вывод по иследованию параметра maxiter
Параметр maxiter задает количество итерации(эпох). По показателю nit видно, что количестов итераций = 154. Поэтому  maxiter = 500-1000 вполне подойдет для нашего случая. Когда maxiter < 154, то выводится информационное сообщение "message: 'Maximum number of iterations has been exceeded.'"

In [60]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = 1.0 ,recombination = 0.7,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 6263
     nit: 154
 success: True
       x: array([0., 0.])
CPU times: user 542 ms, sys: 3.72 ms, total: 546 ms
Wall time: 547 ms


In [61]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = 0.8 ,recombination = 0.7,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 4703
     nit: 115
 success: True
       x: array([0., 0.])
CPU times: user 419 ms, sys: 0 ns, total: 419 ms
Wall time: 417 ms


In [62]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = 0.5 ,recombination = 0.7,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 2503
     nit: 60
 success: True
       x: array([0., 0.])
CPU times: user 225 ms, sys: 0 ns, total: 225 ms
Wall time: 222 ms


In [65]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 0.7,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 3903
     nit: 95
 success: True
       x: array([0., 0.])
CPU times: user 335 ms, sys: 7.68 ms, total: 343 ms
Wall time: 348 ms


#### Вывод по иследованию параметра mutation

Параметр mutation отвечает за силу мутации поколения. Чем больше показатель, тем сложнее сходимость, однако больше шансов найти лучший минимум для функции. По первым трем экпериментам, мы наблюдем этот эффект. В 4 эксперименте, мы задали кортеж из мин и макс значении вероятности мутации, который отбирается в случайном порядке. Производителность алгоритма уходшилась не значительно от лучшео резульатата, однако потенциално увеличивает эффективность поиска экстремума. При этом качество найденого минимума не ухудшилось. 


In [66]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 0.7,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 3903
     nit: 95
 success: True
       x: array([0., 0.])
CPU times: user 336 ms, sys: 9.32 ms, total: 345 ms
Wall time: 345 ms


In [67]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 3423
     nit: 83
 success: True
       x: array([0., 0.])
CPU times: user 305 ms, sys: 0 ns, total: 305 ms
Wall time: 306 ms


In [68]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 0.0,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 5103
     nit: 125
 success: True
       x: array([0., 0.])
CPU times: user 456 ms, sys: 135 µs, total: 456 ms
Wall time: 468 ms


In [69]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 0.3,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 4943
     nit: 121
 success: True
       x: array([0., 0.])
CPU times: user 432 ms, sys: 3.62 ms, total: 436 ms
Wall time: 435 ms


In [70]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 0.5,
                                tol = 0.0000000001,
                                strategy = 'best1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 4583
     nit: 112
 success: True
       x: array([0., 0.])
CPU times: user 409 ms, sys: 0 ns, total: 409 ms
Wall time: 408 ms


#### Вывод по иследованию параметра recombination

Видно, что чем выше вероятность мутации, тем быстрее сходится алгоритм, при этом качество решения не ухудшается.

In [77]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.0000001,
                                strategy = 'best1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 3423
     nit: 83
 success: True
       x: array([0., 0.])
CPU times: user 301 ms, sys: 0 ns, total: 301 ms
Wall time: 304 ms


In [74]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.1,
                                strategy = 'best1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 3423
     nit: 83
 success: True
       x: array([0., 0.])
CPU times: user 306 ms, sys: 56 µs, total: 306 ms
Wall time: 305 ms


#### Вывод по иследованию параметра tol

Качество сходимости наших данных хорошая, поэтому изменение порога допуска на сходимость не влияет на производительность и качество.

In [78]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.1,
                                strategy = 'best1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 3423
     nit: 83
 success: True
       x: array([0., 0.])
CPU times: user 298 ms, sys: 7.76 ms, total: 306 ms
Wall time: 304 ms


In [79]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.1,
                                strategy = 'best1exp')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 3423
     nit: 83
 success: True
       x: array([0., 0.])
CPU times: user 285 ms, sys: 4.03 ms, total: 289 ms
Wall time: 295 ms


In [80]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.1,
                                strategy = 'rand1exp')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 6623
     nit: 163
 success: True
       x: array([0., 0.])
CPU times: user 551 ms, sys: 0 ns, total: 551 ms
Wall time: 551 ms


In [81]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.1,
                                strategy = 'randtobest1exp')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 4023
     nit: 98
 success: True
       x: array([0., 0.])
CPU times: user 374 ms, sys: 37 µs, total: 374 ms
Wall time: 371 ms


In [82]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.1,
                                strategy = 'currenttobest1exp')
print(result)

     fun: 4.440892098500626e-16
 message: 'Maximum number of iterations has been exceeded.'
    nfev: 20103
     nit: 500
 success: False
       x: array([0., 0.])
CPU times: user 2.22 s, sys: 7.65 ms, total: 2.23 s
Wall time: 2.23 s


In [83]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.1,
                                strategy = 'best2exp')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 6063
     nit: 149
 success: True
       x: array([0., 0.])
CPU times: user 514 ms, sys: 0 ns, total: 514 ms
Wall time: 518 ms


In [84]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.1,
                                strategy = 'rand2exp')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 9343
     nit: 231
 success: True
       x: array([0., 0.])
CPU times: user 784 ms, sys: 3.52 ms, total: 788 ms
Wall time: 792 ms


In [85]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.1,
                                strategy = 'randtobest1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 4023
     nit: 98
 success: True
       x: array([0., 0.])
CPU times: user 399 ms, sys: 0 ns, total: 399 ms
Wall time: 398 ms


In [86]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.1,
                                strategy = 'currenttobest1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Maximum number of iterations has been exceeded.'
    nfev: 20103
     nit: 500
 success: False
       x: array([0., 0.])
CPU times: user 2.36 s, sys: 3.63 ms, total: 2.36 s
Wall time: 2.38 s


In [87]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.1,
                                strategy = 'best2bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 6063
     nit: 149
 success: True
       x: array([0., 0.])
CPU times: user 541 ms, sys: 11.7 ms, total: 553 ms
Wall time: 565 ms


In [88]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.1,
                                strategy = 'rand2bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 9343
     nit: 231
 success: True
       x: array([0., 0.])
CPU times: user 823 ms, sys: 0 ns, total: 823 ms
Wall time: 832 ms


In [89]:
%%time
result = differential_evolution(ackley, bounds, seed=42,
                                maxiter = 500,popsize = 20,
                                mutation = (0.5,1.0) ,recombination = 1.0,
                                tol = 0.1,
                                strategy = 'rand1bin')
print(result)

     fun: 4.440892098500626e-16
 message: 'Optimization terminated successfully.'
    nfev: 6623
     nit: 163
 success: True
       x: array([0., 0.])
CPU times: user 576 ms, sys: 7.78 ms, total: 584 ms
Wall time: 596 ms


#### Вывод по иследованию параметра strategy

В документации не описаны ситуации применимости той или иной стратегии. Поэтому, без дополнтельного исследования, нам остается только отнестись к ним как к черному ящику и посмотреть на результаты работы каждой стратегии. 

Лучшими стратеиями, для наших данных, по количеству итераций и времени до сходимости: 'best1bin','best1exp'.
Худшими стратеиями, для наших данных, по количеству итераций и времени до сходимости: 'currenttobest1exp', 'currenttobest1bin'.